### PR TITLE
Fix broken Accept header parsing and content negotiation

### PR DESCRIPTION
**Pull request by kreynolds**
This does not completely fix the accept header parsing but is the least disruptive way to make progress and pave the way towards having proper content negotiation in the future. It has a new Mime::Type.parse method which more strictly adheres to rfc2616 allowing multiple token quoted token types and changes the content_negotiation method to use the Accept header priorities over controller-specific priorities unless the wildcard is used in which case it uses the controller-defined formats.

In a future set of patches the parameter information should be pushed back up the stack into the controller so that its readily available without reparsing the header but thats a much more disruptive patch and I think this is a good place to start.

A few of the tests had to be changed because they were technically incorrect.

### CODE REVIEW

**Comment by joshjordan** [comment]
@kreynolds you did good work here, and there are definitely people that would like to see this merged. Any chance you'll revisit it?

**Comment by jcoyne** [comment]
@rafaelfranca I really want this to go in. What can I do to push this forward?

**Comment by kreynolds** [comment]
I can update this to resolve conflicts if there is a reasonable expectation that it will actually be merged

**Comment by mb21** [comment]
Since the rails-bot assigned this to @rafaelfranca a response from him would be great...

**Comment by iangreenleaf** [comment]
@rafaelfranca any chance of this being accepted? It's a legitimate problem. I think pretty much anyone who has a publicly available web app and keeps a close eye on their error logs has encountered this, since bots send weird stuff that triggers problems with the header parsing.

**Comment by kreynolds** [comment]
The offer to resolve conflicts with rails master to update it still stands

**Comment by rafaelfranca** [comment]
I'm still very uncomfortable with the size of this patch and the fact that it will change behavior for existing applications.

I'd prefer a patch only to change the parser and keep the content negotiation the same and later we could discuss if we should change the content negotiation. Does it make sense?

**Comment by kreynolds** [comment]
I don't believe making a distinction between the two is useful, they are integrated changes and trying to split them up has no real benefit. As mentioned, if I can get a list of current blockers, I will address them and rebase.

**Comment by joshjordan** [comment]
@rafaelfranca please take @kreynolds up on the offer. This does effect existing applications but existing applications all across the web are broken because of it. Spamming logs and disobeying the spec. If the code is safe (eg not producing exceptions in any new case), we really should take this on.

**Comment by rafaelfranca** [comment]
Those are the current blocker: This change is too big and will break existing applications. The only way I see to move this forward is make smaller changes.

**Comment by joshjordan** [comment]
Can you articulate a change that is smaller and still fixes the issue?

**Comment by rafaelfranca** [comment]
The collection of changes to fix the issues we want to address would be probably of the same size, but it would make easier to review and merge if we could break this PR in many steps. I find it really hard to believe we need to change 170 lines in a single commit to be able to fix this issue. We probably could break the first commit in many commits, test the class better and that would give more confidence over what is being changed. Right now I see a huge diff, with a few assertions changes and no real reason why all those lines had to change.

**Comment by kreynolds** [comment]
* #negotiate_mime change is clarification of variable, no functional change
* #valid_accept_header no longer requires a wonky workaround (because its fixed)
* `AcceptItem` can now sort with itself properly (`AcceptList` can go away, this was buggy)
* #parse _used_ to use `AcceptList`, which it can no longer (and doesn't need to, an array of `AcceptItem`s can be sorted properly now). The rest of the changes here are filtering out garbage and mapping `AcceptItem`s to their respective mime types.

The test that is removed is because **_the behavior was incorrect_**. Asking for JSON or any and getting HTML in return when JSON is available is **_wrong_**.

The test for `parse text with trailing star at the beginning` also did not return mimetypes in the correctly specified order, it was **_wrong_**.

I added a new test for a completely broken Accepts header to ensure it defaulted properly.

So in summary, the changes are integrated, no new classes were added, no new methods (except private sorting methods) were added (many were removed), all existing tests pass except for the ones that **_were incorrect_**, and a new test was added to handle invalid data.

I'm not sure why we're hung up on the number of lines changed here, that doesn't seem relevant.

**Comment by joshjordan** [comment]
The tests seem thorough and correct. Are there additional cases that concern you, @rafaelfranca ? I would be up for writing additional tests if there are gaps.

**Comment by joshjordan** [comment]
I don’t understand why we didn’t hear back from the Rails team on this.

**Comment by kaspth** [comment]
We've stated what's required multiple times now: https://github.com/rails/rails/pull/14540#issuecomment-567126332 https://github.com/rails/rails/pull/14540#issuecomment-567129467, and the follow up arguments weren't convincing. Follow Rafael's advice in new smaller PRs to advance this or it'll simply stay closed ✌️

**Comment by joshjordan** [comment]
The latest reply in this thread is a response to Rafael. It gives great
detail on why this PR looks the way it does. It fixed a number of related
bugs that don’t make sense to fix separately. Is there just no time to read
it all?
